packages/@glimmer/syntax/lib/get-template-locals.ts

```diff
@@ -0,0 +1,106 @@
+import { isKeyword } from './keywords';
+import { preprocess } from './parser/tokenizer-event-handlers';
+import traverse from './traversal/traverse';
+import * as ASTv1 from './v1/api';
+
+/**
+ * Gets the correct Token from the Node based on it's type
+ */
+function tokensFromType(node: ASTv1.Node, scopedTokens: string[]): string | void {
+  if (node.type === 'PathExpression') {
+    if (node.head.type === 'AtHead' || node.head.type === 'ThisHead') {
+      return;
+    }
+
+    const possbleToken = node.head.name;
+
+    if (scopedTokens.indexOf(possbleToken) === -1) {
+      return possbleToken;
+    }
+  } else if (node.type === 'ElementNode') {
+    const { tag } = node;
+
+    const char = tag.charAt(0);
+
+    if (char === ':' || char === '@') {
+      return;
+    }
+
+    if (scopedTokens.indexOf(tag) !== -1) {
+      return;
+    }
+
+    return tag;
+  }
+}
+
+/**
+ * Adds tokens to the tokensSet based on their node.type
+ */
+function addTokens(tokensSet: Set<string>, node: ASTv1.Node, scopedTokens: string[]) {
+  const maybeTokens = tokensFromType(node, scopedTokens);
+
+  (Array.isArray(maybeTokens) ? maybeTokens : [maybeTokens]).forEach((maybeToken) => {
+    if (
+      maybeToken !== undefined &&
+      !maybeToken.startsWith('@') &&
+      maybeToken.match(/^[$A-Z_][0-9A-Z_$]*$/i)
+    ) {
+      tokensSet.add(maybeToken);
+    }
+  });
+}
+
+/**
+ * Parses and traverses a given handlebars html template to extract all template locals
+ * referenced that could possible come from the praent scope. Can exclude known keywords
+ * optionally.
+ */
+export function getTemplateLocals(html: string, options?: { excludeKeywords: boolean }): string[] {
+  const ast = preprocess(html);
+  const tokensSet = new Set<string>();
+  const scopedTokens: string[] = [];
+
+  traverse(ast, {
+    Block: {
+      enter({ blockParams }) {
+        blockParams.forEach((param) => {
+          scopedTokens.push(param);
+        });
+      },
+
+      exit({ blockParams }) {
+        blockParams.forEach(() => {
+          scopedTokens.pop();
+        });
+      },
+    },
+
+    ElementNode: {
+      enter(node) {
+        node.blockParams.forEach((param) => {
+          scopedTokens.push(param);
+        });
+        addTokens(tokensSet, node, scopedTokens);
+      },
+
+      exit({ blockParams }) {
+        blockParams.forEach(() => {
+          scopedTokens.pop();
+        });
+      },
+    },
+
+    All(node) {
+      addTokens(tokensSet, node, scopedTokens);
+    },
```

Comment:
Can we move to using `PathExpression`? Are there any other ways to reference a "bound value"?